### PR TITLE
meta: Add in-memory and file based META streams

### PR DESCRIPTION
**Pull request by lmars**
This makes META streams first class citizens in go-meta and managed by the individual stores.

An in-memory META stream just broadcasts writes to all readers using channels.

A file based META stream is a new-line separated list of CIDs stored in a local file, writers append to the end of the file and readers tail the file waiting for new CIDs to be written.

Swarm based streams are not yet implemented as there is no easy way to append to existing data via the API (so for now it just falls back to a local file based stream).

There is a stream HTTP API at `/stream` which supports reading and writing to a META stream. Reads are a stream of Server-Sent-Events, with each event being the object payload, and writes are a POST request to append to the stream:

* write to the "foo.meta" stream:

```
$ curl -X POST -H "Content-Type: application/json" --data-binary '{"value": 1}' http://localhost:5000/stream/foo.meta
{"/":"zdpuAwhdyitq3qhsdhmCepsiWmXHbSnVWSevyuzjtLx6gpZGB"}

$ curl -X POST -H "Content-Type: application/json" --data-binary '{"value": 2}' http://localhost:5000/stream/foo.meta
{"/":"zdpuB2kFRbN7wCw71XetiwTQdJ8xL1sWTPtprapDu5M2vvA9P"}

$ curl -X POST -H "Content-Type: application/json" --data-binary '{"value": 3}' http://localhost:5000/stream/foo.meta
{"/":"zdpuB19ATtxLsSMi2FR8AKbY1CUR3iYDY1PiEuucYmD6Qs8Vn"}
```

* read the stream:

```
$ curl http://localhost:5000/stream/foo.meta
data: {"value":1}

data: {"value":2}

data: {"value":3}
```

### CODE REVIEW

File: stream/api/client.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/flynn/flynn/pkg/httpclient"
+	"github.com/flynn/flynn/pkg/stream"
+	"github.com/ipfs/go-cid"
+)
+
+type Client struct {
+	*httpclient.Client
+}
+
+func NewClient(baseURL string) *Client {
```
Comment (orenyodfat) [inline stream/api/client.go]: Comment public functions and structs

File: stream/file.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+// NewFileStream returns a Stream which reads and writes CIDs to a single file.
+func NewFileStream(path string) Stream {
+	return &fileStream{path: path}
+}
+
+type fileStream struct {
+	path string
+}
+
+func (f *fileStream) NewWriter() (StreamWriter, error) {
```
Comment (orenyodfat) [inline stream/file.go]: Comment public function

File: stream/stream.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import "github.com/ipfs/go-cid"
+
+// Stream provides readers and writers for an underlying META stream.
+type Stream interface {
+	NewReader() (StreamReader, error)
+	NewWriter() (StreamWriter, error)
+}
+
+// StreamWriter writes CIDs to an underlying META stream.
+type StreamWriter interface {
```
Comment (orenyodfat) [inline stream/stream.go]: Consider calling that  Writer (instead StreamWriter).

File: stream/stream.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	Write(...*cid.Cid) error
+	Close() error
+}
+
+// StreamReader reads CIDs from an underlying META stream.
+type StreamReader interface {
```
Comment (orenyodfat) [inline stream/stream.go]: Consider calling that Reader

File: stream/memory.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	m    *memoryStream
+	ch   chan *cid.Cid
+	done chan struct{}
+}
+
+func (m *memoryStreamReader) Ch() <-chan *cid.Cid {
+	return m.ch
+}
+
+func (r *memoryStreamReader) Close() error {
```
Comment (orenyodfat) [inline stream/memory.go]: func receiver should be named "m" (instead of "r") as the previous func.

**Comment by lmars** [comment]
The storage has been changed in https://github.com/meta-network/go-meta/pull/41 so I'm going to close this and re-open with a revised API after that is merged.

**Pull request closed without merging**